### PR TITLE
Allow editing OZW node numeric and list config options

### PR DESCRIPTION
**Pull request by cgarwood**
## Proposed change

Adds support for editing numeric (Int, Byte, Short) and List config items for nodes in the OZW config panel.

image

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- Backend PR: https://github.com/home-assistant/core/pull/43058

## Checklist

- [x] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 1]
```diff
@@ -181,6 +174,98 @@ class OZWNodeConfig extends LitElement {
...
+        ${["Byte", "Short", "Int"].includes(item.type)
+          ? html` <paper-input
+              type="number"
+              .value=${item.value}
+              .max=${item.max}
+              .min=${item.min}
+              ?disabled=${item.parameter === 999}
+              .parameter=${item.parameter}
+            >
+            </paper-input>`
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:191]: Can't we refactor this in the backend so we can just use `ha-form`?

Comment (cgarwood) [reply src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:191]: I'm not familiar with `ha-form` or it's workings, is there a good resource to read up on that?

Comment (bramkragten) [reply src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:191]: Let's talk on discord

**Comment by cgarwood** [comment]
Getting close. A couple questions:

- Is there a way to return an error message back to <ha-form> like you can with the regular paper-input elements?

- Still not sure what to do about `value-changed` not getting fired if the value gets set back to the default (eg going from `50` -> `51` -> `50`, value-changed won't get fired for the last `50`). I tried setting `value` on the <ha-form> element but that didn't seem to have any effect, and from what I can tell `data` on <ha-form> is getting updated automatically.

- Would be nice to disable the slider for config options with a minimum and maximum value, since Z-Wave devices have lots of numeric config values with crazy ranges (0-2678400 for example)

File: src/data/ozw.ts [hunk 2]
```diff
@@ -71,6 +72,7 @@ export interface OZWDeviceConfig {
   min: number;
   max: number;
```
Comment (bramkragten) [inline src/data/ozw.ts:73]: min and max should be part of the schema?

**Comment by bramkragten** [comment]
> Is there a way to return an error message back to like you can with the regular paper-input elements?

Yes, you can pass an `error` and optional a `computeError` function to `ha-form`. It will not show in the input control but above.

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 3]
```diff
@@ -181,6 +180,61 @@ class OZWNodeConfig extends LitElement {
...
+                    <ha-switch ?checked=${option.value} disabled></ha-switch>
+                    ${option.label}
+                  </p>
+                `
+              )}
+              <p>
+                <em>
+                  This configuration option can't be edited from the UI yet
+                </em>
+              </p>
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:211]: This is printed twice now for `BitSet`, also on line 219?

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 4]
```diff
@@ -181,6 +180,61 @@ class OZWNodeConfig extends LitElement {
...
 
+  private _generate_config_block(item) {
+    return html` <ha-card class="content" .value=${item.value}>
+      <div class="card-content">
+        <b>${item.label}</b><br />
+        <span class="secondary">${item.help}</span>
+        ${["Byte", "Short", "Int", "List"].includes(item.type)
+          ? html` <ha-form
+              .schema=${item.schema}
+              .data=${item.data}
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:191]: You should use the value from `this._configData` here

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 5]
```diff
@@ -1,3 +1,8 @@
+import "../../../../../components/ha-switch";
+import "@polymer/paper-item/paper-item";
+import "@polymer/paper-listbox/paper-listbox";
+import "@polymer/paper-input/paper-input";
+import "@polymer/paper-dropdown-menu/paper-dropdown-menu";
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:5]: Not used?

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 6]
```diff
@@ -29,6 +35,8 @@ import {
 import { ERR_NOT_FOUND } from "../../../../../data/websocket_api";
 import { showOZWRefreshNodeDialog } from "./show-dialog-ozw-refresh-node";
 import { ozwNodeTabs } from "./ozw-node-router";
+import { HaForm } from "../../../../../components/ha-form/ha-form";
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:38]:
```suggestion
import type { HaForm } from "../../../../../components/ha-form/ha-form";
```

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 7]
```diff
@@ -29,6 +35,8 @@ import {
...
+import { Button } from "@material/mwc-button/mwc-button";
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:39]:
```suggestion
import type { Button } from "@material/mwc-button/mwc-button";
```

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 8]
```diff
@@ -216,6 +280,28 @@ class OZWNodeConfig extends LitElement {
     }
   }
 
+  private _configValueChanged(ev: CustomEvent) {
+    this._configData[(<OzwHaForm>ev.currentTarget)!.parameter] =
+      ev.detail.value[Object.keys(ev.detail.value)[0]];
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:285]: Make a copy so the form is rerendered
```suggestion
    this._configData = {...this._configData, (<OzwHaForm>ev.currentTarget)!.parameter: ev.detail.value[Object.keys(ev.detail.value)[0]]};
```

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 9]
```diff
@@ -46,6 +54,9 @@ class OZWNodeConfig extends LitElement {
 
   @property() public nodeId?;
 
+  @property()
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:57]:
```suggestion
  @internalProperty()
```

**Comment by cgarwood** [comment]
OK I think it's ready for review again.

Still using sliders for numeric inputs since that's what `<ha-form>` renders, would be nice to be able to use a regular text input but that's probably better as a PR against `<ha-form>`

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 10]
```diff
@@ -181,6 +179,70 @@ class OZWNodeConfig extends LitElement {
     `;
   }
 
+  private _generate_config_block(item) {
+    return html` <ha-card class="content" .value=${item.value}>
+      <div class="card-content">
+        <b>${item.label}</b><br />
+        <span class="secondary">${item.help}</span>
+        ${["Byte", "Short", "Int", "List"].includes(item.type)
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:187]: Can we make this a const?

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 11]
```diff
@@ -216,6 +290,38 @@ class OZWNodeConfig extends LitElement {
     }
   }
 
+  private _configValueChanged(ev: CustomEvent) {
+    this._configData = {
+      ...this._configData,
+      [(<OzwHaForm>ev.currentTarget)!.parameter]: {
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:296]: make this a const

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 12]
```diff
@@ -216,6 +290,38 @@ class OZWNodeConfig extends LitElement {
...
+        [Object.keys(
+          this._configData[(<OzwHaForm>ev.currentTarget)!.parameter]
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:298]: Does this always exist?

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 13]
```diff
@@ -216,6 +290,38 @@ class OZWNodeConfig extends LitElement {
...
+        )[0]]: ev.detail.value[Object.keys(ev.detail.value)[0]],
+      },
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:300]: Not completely sure what is going on here, why can't we just take `ev.detail.value`? I think that has the right key and value already?

File: src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts [hunk 14]
```diff
@@ -216,6 +290,38 @@ class OZWNodeConfig extends LitElement {
...
+    };
+  }
+
+  private async _updateConfigOption(ev: CustomEvent) {
+    const parameter = (<OzwConfigFormButton>ev.currentTarget)!.parameter;
+    const value = this._configData[parameter][
+      Object.keys(this._configData[parameter])[0]
+    ];
```
Comment (bramkragten) [inline src/panels/config/integrations/integration-panels/ozw/ozw-node-config.ts:308]:
```suggestion
    const value = Object.values(this._configData[parameter])[0];
```

**Comment by cgarwood** [comment]
Closing this to focus efforts on the new `zwave_js` integration. Someone else is welcome to pick this up if they would like 🙂

**Pull request closed without merging**